compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -235,6 +235,16 @@ impl<'tcx> ConstToPat<'tcx> {
                             PointerPattern,
                         );
                     }
+                    _ if !self.type_may_have_partial_eq_impl(cv.ty()) => {
```

Comment:
Quoting from the code of that function:
```
        // FIXME: should this call a `predicate_must_hold` variant instead?
        self.infcx.predicate_may_hold(&partial_eq_obligation)
```
@lcnr  any advice on which one we should use? Sadly predicate_must_hold doesn't seem to exist any more, though there are still 2 comments (this one and the doc comment for evaluate_obligation) that refer to predicate_must_hold.

Reply:
:thinking: I think we should use `predicate_must_hold_modulo_regions` here and rename the function to `fn type_has_partial_eq_impl`. The "may have" here isn't what we want given that `ty` is always fully resolved and does not contain inference variables.

I think that `predicate_may_hold` should only differ from `predicate_must_hold_modulo_regions` in cases where MIR borrowck would then ICE because of unexpected ambiguity. I was sadly not able to quickly write a test for that :thinking:

Reply:
"may hold" sounds concerning so I'd prefer "must hold"... but "modulo regions" also sounds concerning, does that mean the PartialEq call we emit later could fail when regions are taken into account?

Reply:
yes, but that is checked by MIR borrowck and results in errors there. There isn't really a good way to check "considering regions" inside of bodies unless you're in MIR borrowck. 

Trait solving during HIR typeck and mir building should generally not be soundness relevant as MIR borrowck should recheck all trait bounds

Reply:
What even is the difference between `predicate_may_hold` and `predicate_must_hold_modulo_regions`? Where does `may_hold` become conservative?

I have found `predicate_must_hold_considering_regions` and that seems like the safest most conservative option, so I figured we could start with that and see if it causes any problems. :)

Reply:
Ah, `predicate_must_hold_considering_regions` is way too conservative, it rejects `&str` even though that type is `PartialEq` for every possible lifetime. I guess it has to be `predicate_must_hold_modulo_regions` then.

Reply:
> Where does `may_hold` return "true" even when it doesn't hold?

in case there is ambiguity. This can happen both because of inference variables, e.g. `?x: Debug` is ambiguous because we don't know whether an arbitrary inference variable implements `Debug`, and because of multiple possible ways to prove something, e.g. normalizing `<T as Iterator>::Item` if there's both a `T: Iterator<Item = u32>` and `T: Iterator<Item = u64>` where-bound.

We also sometimes get ambiguity because our trait solver isn't great™ :sparkles:

Reply:
`predicate_must_hold_modulo_regions` means that a goal definitely holds but proving the returned region constraints may result in a region error. Region constraints are pretty much exclusively checked *outside* of the trait solver and all at once. Proving region constraints requires knowledge of the whole "outlives graph", so we can only do it once we've collected all outlives constraints